lib/internal/readline/utils.js

```diff
@@ -34,13 +34,31 @@ if (internalBinding('config').hasIntl) {
   const icu = internalBinding('icu');
   getStringWidth = function getStringWidth(str, options) {
     options = options || {};
-    if (!Number.isInteger(str))
-      str = stripVTControlCharacters(String(str));
-    return icu.getStringWidth(
-      str,
-      Boolean(options.ambiguousAsFullWidth),
-      Boolean(options.expandEmojiSequence)
-    );
+    if (Number.isInteger(str)) {
+      return icu.getStringWidth(
+        str,
+        Boolean(options.ambiguousAsFullWidth),
+        Boolean(options.expandEmojiSequence)
+      );
+    }
+    str = stripVTControlCharacters(String(str));
+    let width = 0;
+    for (let i = 0; i < str.length; i++) {
+      // Try to avoid calling into C++ by first handling the ASCII portion of
+      // the string. If it is fully ASCII, we skip the C++ part.
+      const code = str.charCodeAt(i);
+      if (code < 127) {
+        width += code >= 32;
```

Comment:
> What if e.g. `code === 9`, a horizontal tab? Reading through `lib/readline.js`, `getStringWidth()` is a bit of a misnomer because it's really used as the display width.

@bnoordhuis That’s a really good point, and I have no good idea for how to handle that – it won’t have a fixed length in any case. This PR doesn’t seem to make anything worse, fwiw.

> Also, I might be mistaken but I think the C++ code considers control characters as being 1 wide, not 0. (Then again, that seems wrong to me.)

It does count control characters as having zero width (and the added test code here should verify that too):

https://github.com/nodejs/node/blob/b3172f834f418f4f2656d851e585f17aece73333/src/node_i18n.cc#L714-L722

Reply:
I did see that code but I couldn't figure out if it actually covers all control characters (`u_charType()` consults a precomputed trie, making it a bit opaque.)

As to the tests you added, those only verify that `getStringWidth()` agrees with itself on the character's vs. the code point's length? Perhaps you could check that the return value is 1 for code points < 128?

Reply:
> As to the tests you added, those only verify that `getStringWidth()` agrees with itself on the character's vs. the code point's length? Perhaps you could check that the return value is 1 for code points < 128?

That, and that the fast path yields the same results as the slow paths.

I’ve added explicit tests that check that control characters are assigned width 0 and all other ASCII characters are assigned width 1.